src/auto-aspect-executor.ts

```diff
@@ -45,33 +49,33 @@ export class AutoAspectExecutor implements OnModuleInit {
             originalFn: any;
             metadata?: unknown;
             aopSymbol: symbol;
-          }[] = this.reflector.get(metadataKey, instance[methodName]);
+          }[] = this.reflector.get(metadataKey, target[methodName]);
           if (!metadataList) {
             return;
           }
 
-          // TODO: Support request scope providers
-          if (!wrapper.isDependencyTreeStatic()) {
-            console.warn(
-              `[${
-                wrapper.instance?.constructor?.name || 'Unknown'
-              }] "@tossteam/nestjs-aop" does not yet support request scope providers`,
-            );
-            return;
-          }
-
           for (const { originalFn, metadata, aopSymbol } of metadataList) {
-            const wrappedMethod = lazyDecorator.wrap({
-              instance,
-              methodName,
-              method: originalFn.bind(instance),
-              metadata,
+            const proxy = new Proxy(target[methodName], {
+              apply: (_, thisArg, args) => {
+                const cached = this.wrappedMethodCache.get(thisArg) || {};
+                if (cached[aopSymbol]?.[methodName]) {
+                  return Reflect.apply(cached[aopSymbol][methodName], lazyDecorator, args);
+                }
+                const wrappedMethod = lazyDecorator.wrap({
+                  instance: thisArg,
+                  methodName,
+                  method: originalFn.bind(thisArg),
+                  metadata,
+                });
+                cached[aopSymbol] ??= {};
+                cached[aopSymbol][methodName] = wrappedMethod;
+                this.wrappedMethodCache.set(thisArg, cached);
+                return Reflect.apply(wrappedMethod, lazyDecorator, args);
```

Comment:
I think thisArg should be included instead of LazyDecorator when executing the wrapped method

Reply:
```typescript
  wrap({ method, metadata: options }: WrapParams<any, BazOptions>) {
    return function baz(this: any, ...args: any[]) {
      console.log(this);                 // 'this' object would be delivered by Reflect.apply()
      const originResult = method(...args);
      const sample = this.sampleService.sample();

      return originResult + ':' + sample + '_' + options.options;
    };
  }
``` 
Consider above situations which library user doesn't return arrow function, but function expression on wrap function. When using arrow function for return value of wrap function, `thisArg` and `lazyDecorator` both works well, because `wrappedMethod` is arrow function and use `this` in lexical scope: the surrounding `lazyDecorator`. But when using function expression for return value, the `wrappedMethod` use `this` binding which is delivered by `Reflect.apply()`. 

So I think using lazyDecorator would be proper.. because library users can choose what func type would be returned by wrap method: between arrow func and func expression.
Any suggestion for better way would be thanks for me!

Reply:
한국인이신거 같아서 한국어로 답해도 괜찮을까여 영어못해서..

wrappedMethod에서 this가 lazy decorator로 평가되기를 기대한다면 arrow function을 쓰는게 말씀처럼 맞아보입니다
하지만 anonymous function 으로 wrapped method를 생성한 경우라면 this가 실행시점의 this로 평가되기를 기대할 것 같아요

lazy decorator를 주입한다면 위와 같은 실수를 커버할 수 있겠지만
의도적으로 실행시점의 this가 바인딩되길 원해하는 경우에 오히려 이슈가 생길 수 있고
실제 anonymous function에서 기대할 수 없는 동작이기에 국룰을 바꾸는 부분들이 트러블슈팅을 어렵게 한다고 개인적으로 생각합니당 🙇

Reply:
아 넵 저도 한국어로 얘기하겠습니다 ㅎㅎ
우선 thisArg로 넘겨주는 instance가 이미 WrapParams의 파라미터에 들어있기 때문에 필요할 때 꺼내 쓸 수 있어서 괜찮다고 생각했고,  wrap함수의 리턴값으로 함수 표현식을 쓸 때와 화살표 함수를 쓸 때 둘 다 일관적으로 동작하는게 좋아보여서 위처럼 구현했습니다. 

> 하지만 anonymous function 으로 wrapped method를 생성한 경우라면 this가 실행시점의 this로 평가되기를 기대할 것 같아요

말씀해주신 부분은 미처 생각하지 못한 부분이네요.. 저렇게 함수 표현식으로 쓸 사용자가 얼마나 있을런지 모르겠지만 저도 this가 실행 시점의 this로 평가되기를 기대할 거라는 생각이 드네요 ㅎㅎ 쓰다가 헷갈릴 수 있을 것 같아 docs에 언급을 하는 편이 좋을 것 같습니다.

그럼 thisArg를 Reflect.apply()에 넘겨주는 걸로 커밋 추가할게요.